yarn-project/aztec-rpc/src/aztec_rpc_server/aztec_rpc_server.ts

```diff
@@ -410,6 +463,30 @@ export class AztecRPCServer implements AztecRPC {
     );
   }
 
+  async #debugLogSimulationError(err: SimulationError) {
+    const functionCallStack = err.getCallStack();
+    const noirCallStack = err.getNoirCallStack();
+
+    // Try to resolve the contract and function names of the stack of failing functions.
+    const stackLines: string[] = [
+      ...(await Promise.all(
+        functionCallStack.map(async ({ contractAddress, functionSelector }) => {
+          const contract = await this.db.getContract(contractAddress);
+          const functionAbi = contract?.functions.find(f => f.selector.equals(functionSelector));
+          return `  at ${contract?.name ?? contractAddress.toString()}.${
+            functionAbi?.name ?? functionSelector.toString()
+          }`;
+        }),
+      )),
+      ...noirCallStack.map(
+        sourceCodeLocation =>
+          `  at ${sourceCodeLocation.filePath}:${sourceCodeLocation.line} '${sourceCodeLocation.locationText}'`,
+      ),
+    ];
+
+    this.log([`Simulation error: ${err.message}`, ...stackLines.reverse()].join('\n'));
+  }
```

Comment:
I'd remove this function in favor of two others:

- An enrichSimulationError method of the RPC server that takes a SimulationError and enrichs its stack trace with the information from the RPC server contract db. It doesn't log anything, just returns a richer error object to be handled by the client however they want.
- A formatNoirStackTrace or formatSimulationError utility (a pure function) that can be used from aztec.js or any other client to format an error, enriched or not, to display it to the user.

This way we're letting the consumer of this error process it however they see fit, while having access to all the debug information we can provide.

Reply:
Heh, just saw that you're already doing something like this in `simulatePublicPart`, where you set the call stack and update the error message of the err object.

Reply:
Yeah I think I'm going to add the enrichment in the RPC server and add a format method in the error class itself